scripts/migrate-tags.sh

```diff
@@ -0,0 +1,137 @@
+
+remote='test'
+version_before_package_rename='0.0.0'
+tag_prefix_before_package_rename="$1"
+tmp_dir='/tmp'
+sed_pattern='s/^v//'
+dry_run=true
+
+print-usage() {
+  cat <<EOF
+
+Migrates tags.
+
+$0 [OPTIONS] PACKAGE_NAME
+
+OPTIONS:
+
+-r REMOTE
+--remote REMOTE
+  Specifies the remote git repo where the tags will be pushed.
+
+-v VERSION_BEFORE_PACKAGE_RENAME
+--version-before-package-rename VERSION_BEFORE_PACKAGE_RENAME
+  The version before the package rename. If package was never renamed, omit this and all tag names will be prepended with the '@metamask/' namespace.
+
+-t TAG_PREFIX_BEFORE_PACKAGE_RENAME
+--tag-prefix-before-package-rename TAG_PREFIX_BEFORE_PACKAGE_RENAME
+  Specifies the tag prefix before the package rename. Defaults to the package name.
+
+-d TMP_DIR
+--tmp-dir TMP_DIR
+  Specifies the temporary directory where the $(git-filter-repo)-applied clone of the original repo is located. Defaults to '/tmp'.
+
+-p SED_PATTERN
+--sed-pattern SED_PATTERN
+  sed pattern for extracting verson numbers from the original repo's tag names. Adjust if the original tag names follow a different naming scheme than 'v0.0.0'.
+
+--no-dry-run
+  If specified, the tags will be created and pushed to the remote repo. Otherwise, the tags and associated release commit hashes will only be printed to stdout.
+EOF
+}
+
+while [[ $# -gt 0 ]]; do
+  key="$1"
+
+  case $key in
+  -h | --help)
+    print-usage
+    exit 0
+    ;;
+  -r | --remote)
+    remote="$2"
+    shift # past argument
+    shift # past value
+    ;;
+  -v | --version-before-package-rename)
+    version_before_package_rename="$2"
+    shift # past argument
+    shift # past value
+    ;;
+  -t | --tag-prefix-before-package-rename)
+    tag_prefix_before_package_rename="$2"
+    shift # past argument
+    shift # past value
+    ;;
+  -d | --tmp-dir)
+    tmp_dir="$2"
+    shift # past argument
+    shift # past value
+    ;;
+  -p | --sed-pattern)
+    sed_pattern="$2"
+    shift # past argument
+    shift # past value
+    ;;
+  --no-dry-run)
+    dry_run=false
+    shift # past argument
+    shift # past value
+    ;;
+  *) # package name
+    package_name="$1"
+    shift # past argument
+    ;;
+  esac
+done
+
+get-tag-commit-pairs() {
+  echo "$(cd $tmp_dir/$package_name && git tag --format="%(refname:short)"$'\t'"%(objectname)")"
+}
+
+get-version-subject-pairs() {
+  while IFS=$'\t' read -r tag commit; do
+    version="$(echo "$tag" | sed "$sed_pattern")"
+    subject="$(cd $tmp_dir/$package_name && git log $commit -n 1 --oneline --format='%H%x09%s' | cut -f2)"
+    echo "$version"$'\t'"$subject"
+  done <<<"$(get-tag-commit-pairs)"
+}
+
+get-version-commit-pairs() {
+  while IFS=$'\t' read -r version subject; do
+    commit="$(git log --oneline --format='%H%x09%s' --grep="^$subject$" | cut -f1)"
+    if [[ $commit == '' ]]; then
+      echo "Could not find commit for version '$version' and subject '$subject'."
+    fi
+    echo "$version"$'\t'"$commit"
```

Comment:
Some thoughts on this:

- It doesn't seem that `git log --grep` matches on subject alone; it seems to match on the entire commit message.
- You seem to be asking for the commit ID but never using it. Instead of using `cut` (which again has the same potential problem as I mentioned earlier) what do you think about just returning the commit ID?
- What happens if this returns more than one result?
- If there are no results, you might want to print to stderr instead of stdout.

Putting this together, what do you think about making a new function:

``` bash
find-commits-matching-subject() {
  local expected_subject="$1"
  while IFS=$'\t' read -r commit actual_subject; do
    if [[ $actual_subject == $expected_subject ]]; then
      echo "$commit"
    fi
  done <<<"$(git log --oneline --format='%H%x09%s' --grep="$expected_subject" --fixed-strings)"
}
```

and then changing the below to:

``` suggestion
  local commits
  local num_commits
  local commits_as_string

  while IFS=$'\t' read -r version subject; do
    commits="$(find-commits-matching-subject "$subject")"
    num_commits="$(echo $commits | wc -l | sed -E 's/^[ ]+//')"
    commits_as_string="$(echo $commits | awk '{ if(FNR == 1) { printf "%s", $0 } else { printf ", %s", $0 } }')"
    if [[ $num_commits -eq 0 ]]; then
      echo "Could not find commit for version '$version' and subject '$subject'." >&2
    elif [[ $num_commits -gt 1 ]]; then
      echo "More than one commit found for '$version' and subject '$subject': $commits_as_string" >&2
    else
      echo "$version"$'\t'"$commits"
    fi

Reply:
The commit message search is intended to look for a full match due to https://github.com/MetaMask/core/pull/1802#discussion_r1378298667. Applied all of other the suggestions, thanks for the comprehensive rewrites here!